src/material/menu/menu-trigger.ts

```diff
@@ -519,6 +519,7 @@ export class MatMenuTrigger implements AfterContentInit, OnDestroy {
   /** Handles key presses on the trigger. */
   _handleKeydown(event: KeyboardEvent): void {
     const keyCode = event.keyCode;
+    this._openedBy = 'keyboard';
```

Comment:
I think this change alone could potentially cause a very subtle bug that we may not care about. It's possible for this method to be called and not actually open the menu. For example, if focus is on the menu trigger button and one presses some key that is not Enter or Space. In that case, `this_openedBy` will be set to `keyboard`, even though the menu was not actually opened by keyboard.

This means that a subsequent programmatic call to open the menu would cause this line (`this.menu.focusFirstItem(this._openedBy || 'program');`) within `_initMenu` to focus the first menu item with `keyboard` origin, even though we'd expect for it to be opened with `program` origin.

Again, I think this is a very minor bug and maybe something we don't care about. But one potential fix is to conditionally set `this._openedBy = 'keyboard';` based upon if the menu is actually opened. I'm not sure if it's possible to know whether or not the menu will be opened in this method, but we could use a heuristic (e.g. if Enter or Space are pressed or if the submenu conditions below are true).

Reply:
I think that's very unlikely to happen, because we set the value to something different next time it is opened. I've added an extra check so we only set this when the Enter key is pressed.

Reply:
SGTM. You might want to check to Space as well. Also, I think we also want to set it to `keyboard` when opening a sub menu with arrow keys?